stacks-signer/src/runloop.rs

```diff
@@ -68,60 +50,47 @@ pub struct RunLoop<C> {
     pub signing_round: SigningRound,
     /// The stacks client
     pub stacks_client: StacksClient,
-    /// The DKG Command to perform
-    pub command: RunLoopCommand,
+    /// Received Commands that need to be processed
+    pub commands: VecDeque<RunLoopCommand>,
     /// The current state
     pub state: State,
 }
 
 impl<C: Coordinatable> RunLoop<C> {
     /// Helper function to check the current state and execute commands accordingly
     /// This will update the state if a command is executed
-    fn update_state(&mut self) {
-        match (&self.command, &self.state) {
-            (RunLoopCommand::Run, State::Idle) => {
-                info!("Starting runloop");
-                self.state = State::Run;
-            }
-            (RunLoopCommand::Dkg, State::Idle) => {
-                info!("Starting DKG");
-                if let Ok(msg) = self.coordinator.start_distributed_key_generation() {
-                    let ack = self
-                        .stacks_client
-                        .send_message(self.signing_round.signer.signer_id, msg);
-                    debug!("ACK: {:?}", ack);
-                }
-                self.state = State::Dkg;
-            }
-            (RunLoopCommand::Dkg, State::DkgDone) => {
-                info!("Done DKG");
-                self.state = State::Exit;
-            }
-            (RunLoopCommand::DkgSign { message }, State::Idle) => {
-                info!("Starting DKG for signing message: {:?}", message);
-                if let Ok(msg) = self.coordinator.start_distributed_key_generation() {
-                    let _ = self
-                        .stacks_client
-                        .send_message(self.signing_round.signer.signer_id, msg);
-                }
-                self.state = State::Dkg;
-            }
-            (RunLoopCommand::DkgSign { message }, State::DkgDone)
-            | (RunLoopCommand::Sign { message }, State::Idle) => {
-                info!("Signing message: {:?}", message);
-                if let Ok(msg) = self.coordinator.start_signing_message(message) {
-                    let _ = self
-                        .stacks_client
-                        .send_message(self.signing_round.signer.signer_id, msg);
+    fn execute_command(&mut self) {
+        match self.state {
+            State::Idle => {
+                if let Some(command) = self.commands.pop_front() {
+                    match command {
+                        RunLoopCommand::Dkg => {
+                            info!("Starting DKG");
+                            if let Ok(msg) = self.coordinator.start_distributed_key_generation() {
```

Comment:
What happens if this fails? What's the `Err(..)` case?  Same question for `RunLoopCommand::Sign` below with `start_signing_message()`.

Reply:
Good point. If it fails, we should not update the state to State::Dkg because it will become stuck. We should move the state update internal if let Ok() check. Currently if start_distributed_key_generation fails. Beyond this, it would imply that there is something substantially wrong with the coordinator state. Not sure if we should panic here, just warn. Attempt to push back the command and reprocess it? Not sure how it should fail here. @xoloki

Reply:
Done :)